### __Group Assignment - Predicting Airbnb Listing Prices in Melbourne, Australia__

--- 

**Kaggle Competition Ends:** Friday, 6 June 2025 @ 3:00pm (Week 13)  
**Assignment Due Date on iLearn:** Friday, 6 June 2025 @ 11.55pm (Week 13)   
**Total Marks:** 30

**Overview:**   

- In the group assignment you will form a team of 3 students and participate in a forecasting competition on Kaggle


**Instructions:** 

- Form a team of 3 students 
- Each team member needs to join [https://www.kaggle.com](https://www.kaggle.com/)  
- Choose a team leader and form a team on Kaggle [https://www.kaggle.com/t/fc5974a56165cea945ee1ec182b079af](https://www.kaggle.com/t/fc5974a56165cea945ee1ec182b079af)
    - Team leader to click on `team` and invite other 2 team members to join
    - Your **team's name must start** with our unit code
- All team members should work on all the tasks however   
    - Each team member will be responsible for one of the 3 tasks listed below    
- **Your predictions must be generated by a model you develop here** 
    - You will receive a mark of **zero** if your code is not able produce the forecasts you submit to Kaggle 

**Competition Rankings**

The rankings for the competition are determined through two different leaderboards:

- **Public Leaderboard Ranking**: Available during the competition, these rankings are calculated based on 50% of the test dataset, which includes 1,500 observations. This allows participants to see how they are performing while the competition is still ongoing.
- **Final Leaderboard Ranking**: These rankings are recalculated from the other 50% of the test dataset, which consists of the remaining 1,500 observations, and are revealed 5 minutes after the competition concludes. This final evaluation determines the ultimate standings of the competition.



**Marks** 

- Assignment: 30 marks consisting of Solutions (27 marks) + Video Presentation (3 marks)
- **Each Student's Mark: 50% x overall assignment mark + 50% x mark for the task that you are responsible for**  



**Submissions:**  

1. On Kaggle: submit your team's forecast in order to be ranked by Kaggle
2. On iLearn **only team leader to submit** the assignment Jupyter notebook re-named to your team's name on Kaggle   
    - The Jupyter notebook must contain team members names/ID numbers, and the group name Kaggle
    - One 15 minute video recording of your work 
        - 5 marks will be deducted from each Task for which there is no video presentation   


---
---

### <span style="background-color: yellow;">**Fill out the following information**

- Team Name on Kaggle: `(insert here)`
- Team Leader and Team Member 1: `(insert here)`
- Team Member 2: `(insert here)`
- Team Member 3: `(insert here)`

---

## Task 1: Problem Description and Initial Data Analysis

- You must clearly explain all your answers in both the Markdown file and the recorded video.

**Total Marks: 9**   

Based on the Competition Overview, datasets and additional information provided on Kaggle, along with insights gained from personal research of the topic, write **Problem Description** (about 500 words) focusing on the sections listed below: 
1. Forecasting Problem - explain what we are trying to do and how it could be used in the real world, e.g. who and how may benefit from it (2 marks)    
2. Evaluation Criteria - discuss the criterion that is used in this competition to assess forecasting performance, and its pros and cons. (2 marks)     
3. Categorise all variables provided in the dataset according to their type; Hint: similar to what we had in Programming Task 1 (2 marks)  
4. Missing Values - explain what you find for both the training and test datasets at this stage (2 marks)
5. Provide and discuss some interesting *univariate* data characteristics (e.g. summary statistics and plots) in the training dataset  (1 marks)       
- Hints:
    - You should **not** discuss any specific predictive algorithms at this stage


Student in charge of this task: `(insert name here)`

#Task 1 code here, insert more cells if required

## Task 1.1: Forecasting Problem – Explanation and Real-World Application

 ### Overview

   This competition presents a **supervised regression problem**, where the objective is to predict a **continuous target 
   variable**—the `price` of Airbnb listings—based on various input features. These features may include property 
   characteristics such as room type, location, availability, and number of reviews.

### Real-World Relevance

 Accurately forecasting Airbnb prices carries significant value for multiple stakeholders:

  #### Hosts
  Hosts can leverage price prediction models to **set competitive and optimal prices** for their properties. This can lead to 
  **increased occupancy rates** and **higher revenue**, particularly in dynamic markets where pricing strategies need to 
  adjust based on demand and seasonality.

  #### Guests
  Guests benefit from such models through **improved price transparency** and **comparison tools** that help them identify 
  listings offering better value for money. This enhances the user experience and builds trust in the platform.

  ####  Platform Operators (e.g., Airbnb)
  For the platform itself, predictive pricing supports **automated dynamic pricing algorithms**. These systems adjust listing 
  prices in real time based on factors such as:
  - Market demand fluctuations
  - Seasonal patterns
  - Local events
  - Competitor pricing

## Task 1.2: Evaluation Criteria – Mean Absolute Error (MAE)

### Metric Used

The evaluation metric used in this competition is **Mean Absolute Error (MAE)**.  
MAE measures the average magnitude of the errors between predicted and actual values, without considering their direction.

### Advantages of MAE
- ✅ **Easy to interpret**: The error is in the same unit as the target variable (e.g., dollars).
- ✅ **Robust to outliers**: Unlike metrics that square the error, MAE treats all errors linearly and does not give extra weight to large deviations.
- ✅ **Intuitive for business stakeholders**: It directly reflects the average error in predictions.

### Limitations of MAE
- ❌ **Does not penalize large errors** as harshly as other metrics like MSE (Mean Squared Error), which may be important in some contexts.
- ❌ **Non-differentiable at zero**: While not a concern for this task, it can be less convenient for certain optimization algorithms in advanced model training.

## Task 1.3: Variable Categorization

| Variable                                      | Type        |
|-----------------------------------------------|-------------|
| ID                                            | Categorical (Identifier) |
| source                                        | Categorical (Nominal) |
| name                                          | Text        |
| description                                   | Text        |
| neighborhood_overview                         | Text        |
| host_name                                     | Text        |
| host_since                                    | Date        |
| host_location                                 | Text        |
| host_about                                    | Text        |
| host_response_time                            | Categorical (Ordinal) |
| host_response_rate                            | Numeric (%) |
| host_acceptance_rate                          | Numeric (%) |
| host_is_superhost                             | Boolean     |
| host_neighbourhood                            | Text        |
| host_listings_count                           | Numeric     |
| host_verifications                            | Text/JSON   |
| host_has_profile_pic                          | Boolean     |
| host_identity_verified                        | Boolean     |
| neighbourhood                                 | Text        |
| neighbourhood_cleansed                        | Text        |
| latitude                                      | Numeric     |
| longitude                                     | Numeric     |
| property_type                                 | Categorical (Nominal) |
| room_type                                     | Categorical (Nominal) |
| accommodates                                  | Numeric     |
| bathrooms                                     | Text/Numeric |
| bedrooms                                      | Numeric     |
| beds                                          | Numeric     |
| amenities                                     | JSON/Text   |
| minimum_nights                                | Numeric     |
| maximum_nights                                | Numeric     |
| minimum_minimum_nights                        | Numeric     |
| maximum_minimum_nights                        | Numeric     |
| minimum_maximum_nights                        | Numeric     |
| maximum_maximum_nights                        | Numeric     |
| minimum_nights_avg_ntm                        | Numeric     |
| maximum_nights_avg_ntm                        | Numeric     |
| has_availability                              | Boolean     |
| availability_30                               | Numeric     |
| availability_60                               | Numeric     |
| availability_90                               | Numeric     |
| availability_365                              | Numeric     |
| number_of_reviews                             | Numeric     |
| number_of_reviews_ltm                         | Numeric     |
| number_of_reviews_l30d                        | Numeric     |
| first_review                                  | Date        |
| last_review                                   | Date        |
| review_scores_rating                          | Numeric     |
| review_scores_accuracy                        | Numeric     |
| review_scores_cleanliness                     | Numeric     |
| review_scores_checkin                         | Numeric     |
| review_scores_communication                   | Numeric     |
| review_scores_location                        | Numeric     |
| review_scores_value                           | Numeric     |
| instant_bookable                              | Boolean     |
| calculated_host_listings_count                | Numeric     |
| calculated_host_listings_count_entire_homes   | Numeric     |
| calculated_host_listings_count_private_rooms  | Numeric     |
| calculated_host_listings_count_shared_rooms   | Numeric     |
| reviews_per_month                             | Numeric     |
| price                                         | Numeric (Target) |



In [38]:
import pandas as pd

# Load CSV files
train_df = pd.read_csv('train.csv')
test_df = pd.read_csv('test.csv')

# Check if they loaded properly
train_df.head()

# Show count of missing values in each column of the training dataset
missing_train = train_df.isnull().sum()
missing_train = missing_train[missing_train > 0].sort_values(ascending=False)

print("🔹 Missing Values in train.csv:\n")
print(missing_train)

# Show count of missing values in each column of the test dataset
missing_test = test_df.isnull().sum()
missing_test = missing_test[missing_test > 0].sort_values(ascending=False)

print("🔹 Missing Values in test.csv:\n")
print(missing_test)


🔹 Missing Values in train.csv:

host_neighbourhood             3460
host_about                     2270
neighbourhood                  1823
neighborhood_overview          1823
host_location                  1280
bedrooms                        378
room_type                       134
neighbourhood_cleansed          108
property_type                    84
beds                             72
host_acceptance_rate             63
description                      57
availability_365                 47
minimum_minimum_nights           45
maximum_maximum_nights           45
review_scores_cleanliness        22
review_scores_checkin            22
review_scores_value              22
review_scores_accuracy           22
review_scores_location           22
review_scores_communication      22
bathrooms                         6
name                              1
dtype: int64
🔹 Missing Values in test.csv:

host_neighbourhood             2066
host_about                     1441
neighbourhood           

## Task 1.4: Missing Values – Training and Test Datasets

### 🔹 Overview

Analyzing missing data is a crucial step in data preprocessing. It helps identify potential issues with data quality and guides decisions about cleaning, imputing, or discarding variables. The presence of missing values can affect the model’s ability to learn patterns effectively if not handled appropriately.

---

### 🔹 Training Dataset (`train.csv`)

Several columns in the training dataset contain missing values. The most affected columns are:

- `host_neighbourhood` – 3,460 missing values
- `host_about` – 2,270 missing
- `neighbourhood` and `neighborhood_overview` – 1,823 each
- `host_location` – 1,280 missing
- Minor missingness: `bathrooms` (6), `name` (1)

These missing values appear across **text, categorical, and numerical** variables. 

---

### 🔹 Test Dataset (`test.csv`)

Missing values are also observed in similar fields:

- `host_neighbourhood` – 2,066 missing
- `host_about` – 1,441
- `neighbourhood` and `neighborhood_overview` – 1,424 each
- `host_location`, `host_response_time`, and `host_response_rate` – 700+ missing values
- Minor missingness in `bedrooms`, `room_type`, `description`, and `availability_365`

Compared to the training set, the test set has slightly fewer missing values in some fields but introduces additional gaps like `review_scores_rating`, `first_review`, and `last_review`.

---





---

## Task 2: Data Cleaning, Missing Observations and Feature Engineering

- You must clearly explain all your answers in both the Markdown file and the recorded video. 

**Total Marks: 9**

Student in charge of this task: `(insert name here)`

**Task 2, Question 1**: Clean **all** numerical features so that they can be used in training algorithms. For instance, `host_response_rate` feature is in object format containing both numerical values and text. Extract numerical values (or equivalently eliminate the text) so that the numerical values can be used as a regular feature.  
(2 marks)

In [ ]:
## Task 2, Question 1 Code Here

`(Task 2, Question 1 Text Here - insert more cells as required)`

**Task 2, Question 2** Create at least 4 new features from existing features which contain multiple items of information.   
(2 marks)

In [ ]:
## Task 2, Question 2 Code Here

`(Task 2, Question 2 Text Here - insert more cells as required)`

**Task 2, Question 3**: Impute the missing values for all features in both the training and test datasets.   
(2 marks)

In [ ]:
## Task 2, Question 3 Code Here

`(Task 2, Question 3 Text Here - insert more cells as required)`

**Task 2, Question 4**: Encode all categorical variables appropriately as discussed in class. 

- Where multiple values are given for an observation encode the observation as 'other'. 
- Where a categorical feature contains more than 5 unique values, map the features into 5 most frequent values + 'other' and then encode appropriately.  
(2 marks)

In [ ]:
## Task 2, Question 4 Code Here

`(Task 2, Question 4 Text Here - insert more cells as required)`

**Task 2, Question 5**: Perform any additional data preparation steps you consider necessary before building your predictive models, and clearly explain each action you take.  
(1 mark)

In [ ]:
## Task 2, Question 5 Code Here

`(Task 2, Question 5 Text Here - insert more cells as required)`

--- 
## Task 3: Fit and tune predictive models, submit predictions & win competition

- You must clearly explain all your answers in both the Markdown file and the recorded video.
- 
**Total Marks: 9**

For this task, you should not create any new features and must rely on the variables constructed in Task 2.  
 

1. Perform some EDA to measure the relationship between the features and the target variable, and carefully explain your findings. (2 marks)

2. Choose and carefully explain 3 different machine learning (ML) regression models that you will apply in this competition. (2 marks)
   
3. Train the models from the above question and tune their hyperparameters via cross-validation. Discuss the fitted weights, optimised hyperparameter values, and their training dataset predictive performance. (2 marks)   

4. Select your best model, create predictions of the test dataset and submit your forecasts on Kaggle's competition page. Provide Kaggle ranking and score (screenshots) and comment on your performance in the competition. (2 marks)

5. Suggest ways to improve your ranking and implement them, providing further evidence from Kaggle (screenshots). (1 mark)   

- Hints:
    - Make sure your Python code works so that your results can be replicated by the marker
    - You will receive the mark of zero for this Task if your code does not produce the forecasts uploaded to Kaggle



Student in charge of this task: `(insert name here)`

In [ ]:
#Task 3 code here

`(Task 3 - insert more cells as required)`

---
---
## Marking Criteria

To receive full marks your solutions must satisfy the following criteria:

- Problem Description: 9 marks
- Data Cleaning: 9 marks
- Building Forecasting models: 9 marks
- Video Presentation: 3 marks
    - Duration less than 15 min, presentation skill and content 
    - Each team member delivers a 5-minute presentation on their assigned task
    - All assignment questions must be discussed on video  
    - Your code must be readable on the video
    - Discuss both the actions you took and, more importantly, the reasoning behind these actions, explaining the significance of key steps
    - During the video recording, make sure that both your face and Jupyter Notebook are clearly visible
- Forecasts correctly uploaded to Kaggle
- Python code is clean and concise
- Written explanations are provided in clear and easy to understand sentences
- The assignment notebook is well-organised and easy to follow
- Failure to meet the above marking criteria will result in a deduction of marks

---
---